In [1]:
import pandas as pd
import numpy as np

pd.set_option('display.max_columns', None)
pd.set_option('display.float_format', '{:.2f}'.format)

print("Libraries loaded")

Libraries loaded


In [2]:
# ── Dataset 1: Patient Admissions ──────────────────────────────
df_admissions = pd.read_csv('../data/raw/healthcare_dataset.csv')

print("=== SHAPE ===")
print(df_admissions.shape)

print("\n=== COLUMNS & TYPES ===")
print(df_admissions.dtypes)

print("\n=== FIRST 3 ROWS ===")
df_admissions.head(3)

=== SHAPE ===
(55500, 15)

=== COLUMNS & TYPES ===
Name                   object
Age                     int64
Gender                 object
Blood Type             object
Medical Condition      object
Date of Admission      object
Doctor                 object
Hospital               object
Insurance Provider     object
Billing Amount        float64
Room Number             int64
Admission Type         object
Discharge Date         object
Medication             object
Test Results           object
dtype: object

=== FIRST 3 ROWS ===


,Name,Age,Gender,Blood Type,Medical Condition,Date of Admission,Doctor,Hospital,Insurance Provider,Billing Amount,Room Number,Admission Type,Discharge Date,Medication,Test Results
0,Bobby JacksOn,30,Male,B-,Cancer,2024-01-31,Matthew Smith,Sons and Miller,Blue Cross,18856.28,328,Urgent,2024-02-02,Paracetamol,Normal
1,LesLie TErRy,62,Male,A+,Obesity,2019-08-20,Samantha Davies,Kim Inc,Medicare,33643.33,265,Emergency,2019-08-26,Ibuprofen,Inconclusive
2,DaNnY sMitH,76,Female,A-,Obesity,2022-09-22,Tiffany Mitchell,Cook PLC,Aetna,27955.10,205,Emergency,2022-10-07,Aspirin,Normal


In [3]:
print("=== NULL VALUES ===")
print(df_admissions.isnull().sum())

print("\n=== UNIQUE VALUE COUNTS ===")
for col in df_admissions.columns:
    print(f"{col}: {df_admissions[col].nunique()} unique values")

=== NULL VALUES ===
Name                  0
Age                   0
Gender                0
Blood Type            0
Medical Condition     0
Date of Admission     0
Doctor                0
Hospital              0
Insurance Provider    0
Billing Amount        0
Room Number           0
Admission Type        0
Discharge Date        0
Medication            0
Test Results          0
dtype: int64

=== UNIQUE VALUE COUNTS ===
Name: 49992 unique values
Age: 77 unique values
Gender: 2 unique values
Blood Type: 8 unique values
Medical Condition: 6 unique values
Date of Admission: 1827 unique values
Doctor: 40341 unique values
Hospital: 39876 unique values
Insurance Provider: 5 unique values
Billing Amount: 50000 unique values
Room Number: 400 unique values
Admission Type: 3 unique values
Discharge Date: 1856 unique values
Medication: 5 unique values
Test Results: 3 unique values


### Data Quality Check Summary

The dataset contains 55,500 patient admission records and 15 columns.  
The null-value check shows that there are no missing values in any column, so no imputation or row removal is required at this stage.

The unique-value count helps identify the nature of each column. Columns such as Gender, Blood Type, Medical Condition, Insurance Provider, Admission Type, Medication, and Test Results have a small number of unique values, which makes them useful categorical fields for grouping and dashboard analysis.

Columns such as Name, Doctor, and Hospital have a high number of unique values. The Billing Amount column  has many unique values, which is expected because it is a continuous numeric field.

In [4]:
print("=== MEDICAL CONDITIONS ===")
print(df_admissions['Medical Condition'].value_counts())

print("\n=== ADMISSION TYPES ===")
print(df_admissions['Admission Type'].value_counts())

print("\n=== TEST RESULTS ===")
print(df_admissions['Test Results'].value_counts())

print("\n=== INSURANCE PROVIDERS ===")
print(df_admissions['Insurance Provider'].value_counts())

=== MEDICAL CONDITIONS ===
Medical Condition
Arthritis       9308
Diabetes        9304
Hypertension    9245
Obesity         9231
Cancer          9227
Asthma          9185
Name: count, dtype: int64

=== ADMISSION TYPES ===
Admission Type
Elective     18655
Urgent       18576
Emergency    18269
Name: count, dtype: int64

=== TEST RESULTS ===
Test Results
Abnormal        18627
Normal          18517
Inconclusive    18356
Name: count, dtype: int64

=== INSURANCE PROVIDERS ===
Insurance Provider
Cigna               11249
Medicare            11154
UnitedHealthcare    11125
Blue Cross          11059
Aetna               10913
Name: count, dtype: int64


### Categorical Distribution Summary

The categorical distribution check shows that the main healthcare fields are generally balanced. Medical conditions are distributed almost evenly across Arthritis, Diabetes, Hypertension, Obesity, Cancer, and Asthma. This balance makes the dataset useful for comparing admission patterns and billing behaviour across different condition groups.

Admission types are also evenly distributed across Elective, Urgent, and Emergency visits. This allows fair comparison between planned admissions and more urgent hospital visits.

Test results are split almost equally across Abnormal, Normal, and Inconclusive outcomes. This can support later analysis of test-result trends by condition, admission type, medication, or insurance provider.

Insurance providers are also relatively balanced, with Cigna, Medicare, UnitedHealthcare, Blue Cross, and Aetna all having similar record counts. 

In [6]:
# ── Dataset 2: ER Wait Times ──────────────────────────────
df_er = pd.read_csv('../data/raw/ER Wait Time Dataset.csv')

print("=== SHAPE ===")
print(df_er.shape)

print("\n=== COLUMNS & TYPES ===")
print(df_er.dtypes)

print("\n=== NULL VALUES ===")
print(df_er.isnull().sum())

print("\n=== URGENCY LEVELS ===")
print(df_er['Urgency Level'].value_counts())

print("\n=== REGIONS ===")
print(df_er['Region'].value_counts())

print("\n=== PATIENT OUTCOMES ===")
print(df_er['Patient Outcome'].value_counts())

print("\n=== WAIT TIME STATS (minutes) ===")
print(df_er['Total Wait Time (min)'].describe())

=== SHAPE ===
(5000, 19)

=== COLUMNS & TYPES ===
Visit ID                              object
Patient ID                            object
Hospital ID                           object
Hospital Name                         object
Region                                object
Visit Date                            object
Day of Week                           object
Season                                object
Time of Day                           object
Urgency Level                         object
Nurse-to-Patient Ratio                 int64
Specialist Availability                int64
Facility Size (Beds)                   int64
Time to Registration (min)             int64
Time to Triage (min)                   int64
Time to Medical Professional (min)     int64
Total Wait Time (min)                  int64
Patient Outcome                       object
Patient Satisfaction                   int64
dtype: object

=== NULL VALUES ===
Visit ID                              0
Patient ID          

### ER Wait-Time Dataset Summary

The ER wait-time dataset contains 5,000 emergency room visit records and 19 columns. The initial quality check shows that there are no missing values across any of the columns.

The urgency levels are well balanced across Low, Medium, High, and Critical cases, which allows fair comparison of wait-time patterns by urgency. The dataset contains more Urban visits than Rural visits, making region an important variable for analyzing differences in healthcare access and patient flow.

Most patients were discharged, followed by admitted patients. A smaller group left without being seen, which is an important operational outcome because it may indicate long wait times, overcrowding, or access challenges.

The average total wait time is approximately 82 minutes, while the median wait time is 60 minutes. Since the average is higher than the median, the wait-time distribution appears right-skewed, meaning that some very long waits are increasing the overall average.

In [8]:
# ── Dataset 3: CIHI Provincial Wait Times ──────────────────────────────
df_cihi = pd.read_excel('../data/raw/wait-times-priority-procedures-in-canada-2025-data-tables-en.xlsx',
                        sheet_name='Table 1',
                        header=1)

# Drop the empty trailing columns
df_cihi = df_cihi.dropna(axis=1, how='all')

print("=== SHAPE ===")
print(df_cihi.shape)

print("\n=== COLUMNS ===")
print(df_cihi.columns.tolist())

print("\n=== PROVINCES ===")
print(df_cihi['Province'].unique())

print("\n=== PROCEDURES ===")
print(df_cihi['Indicator'].unique())

print("\n=== METRICS ===")
print(df_cihi['Metric'].unique())

print("\n=== YEAR RANGE ===")
print(sorted(df_cihi['Data year'].dropna().astype(str).unique()))
print("\n=== SAMPLE ROWS ===")
df_cihi.head(5)

=== SHAPE ===
(18965, 10)

=== COLUMNS ===
['Reporting level', 'Province', 'Region', 'Indicator', 'Metric', 'Data year', 'Unit of measurement', 'Indicator result', 'Unnamed: 9', 'Unnamed: 11']

=== PROVINCES ===
['Alberta' 'British Columbia' 'Canada' 'Manitoba' 'New Brunswick'
 'Newfoundland and Labrador' 'Nova Scotia' 'Ontario'
 'Prince Edward Island' 'Quebec' 'Saskatchewan' nan]

=== PROCEDURES ===
['Bladder Cancer Surgery' 'Breast Cancer Surgery' 'CABG'
 'Cataract Surgery' 'Colorectal Cancer Surgery' 'CT Scan'
 'Hip Fracture Repair' 'Hip Fracture Repair/Emergency and Inpatient'
 'Hip Replacement' 'Knee Replacement' 'Lung Cancer Surgery' 'MRI Scan'
 'Prostate Cancer Surgery' 'Radiation Therapy' nan]

=== METRICS ===
['50th Percentile' '90th Percentile' 'Volume' '% Meeting Benchmark' nan]

=== YEAR RANGE ===
['2008', '2009', '2010', '2011', '2012', '2013', '2014', '2015', '2016', '2017', '2018', '2019', '2019FY', '2019Q3Q4', '2020', '2020FY', '2020Q3Q4', '2021', '2021FY', '2021Q3Q4', 

,Reporting level,Province,Region,Indicator,Metric,Data year,Unit of measurement,Indicator result,Unnamed: 9,Unnamed: 11
0,Provincial,Alberta,NaN,Bladder Cancer Surgery,50th Percentile,2008,Days,NaN,NaN,NaN
1,Provincial,Alberta,NaN,Bladder Cancer Surgery,90th Percentile,2008,Days,NaN,NaN,NaN
2,Provincial,Alberta,NaN,Bladder Cancer Surgery,Volume,2008,Number of cases,NaN,NaN,NaN
3,Provincial,Alberta,NaN,Breast Cancer Surgery,50th Percentile,2008,Days,NaN,NaN,NaN
4,Provincial,Alberta,NaN,Breast Cancer Surgery,90th Percentile,2008,Days,NaN,NaN,NaN


### CIHI Provincial Wait Times Dataset Summary

The CIHI wait-times dataset contains 18,965 rows and includes provincial, regional, and Canada-level wait-time results for priority procedures. The dataset includes procedures such as cancer surgeries, cataract surgery, CT scans, MRI scans, hip replacement, knee replacement, and radiation therapy.

The `Metric` column is especially important because the meaning of `Indicator result` changes depending on the selected metric. For example, when the metric is `50th Percentile` or `90th Percentile`, the result represents wait time in days. When the metric is `Volume`, the result represents the number of cases. When the metric is `% Meeting Benchmark`, the result represents benchmark performance.

Some extra `Unnamed` columns appeared because the source is a formatted Excel workbook. These columns should be removed before further analysis. The dataset also contains some missing province, indicator, and metric values, which are likely blank or note rows from the Excel file and should be cleaned before analysis.

<h1 style="color:red;">Overall Dataset Summary</h1>

| Dataset | Rows | Columns | Nulls | Key Issues to Fix |
|---|---|---|---|---|
| Patient Admissions | 55,500 | 15 | 0 | Date columns are strings, inconsistent name casing |
| ER Wait Times | 5,000 | 19 | 0 | Visit Date is string, column names are too verbose |
| CIHI Wait Times | 18,965 | 10 | Yes | Mixed year formats (FY, Q3Q4), unnamed columns, mixed data types in year column |